File: src/game_loop.lua
--[[
    GameLoop — ISO with legacy JS GameLoop.js.

    Flow: startRun→IDLE → spin/resolveBallAt→SPINNING → RESULTS →
          continueFromResults→CHOICE → makeChoice/skipChoice→SHOP →
          endShop→IDLE (next round)

    Scoring: baseVal = setBaseValue or (segIndex+1), even/odd bonuses,
             ×weight, +value_plus_2 stacks, gold×2, special ball effects.
    Score carries over: endShop deducts quota, surplus persists.
]]

local BALANCE, getQuota = (function() local b = require('src.data.balance'); return b.BALANCE, b.getQuota end)()
local StateMod = require('src.state')
local PHASE    = StateMod.PHASE

local RNG      = require('src.rng')
local EventBus = require('src.event_bus')

local Scoring  = require('src.systems.scoring')
local WheelS   = require('src.systems.wheel')
local Shop     = require('src.systems.shop')
local ChoiceS  = require('src.systems.choice')
local MetaMod  = require('src.systems.meta')
local Meta     = MetaMod.Meta
local Effect   = require('src.systems.effect')
local getSymbol = require('src.data.symbols').getSymbol

local GL = {}
GL.__index = GL

function GL.new(opts)
    opts = opts or {}
    local self = setmetatable({}, GL)
    self.events  = EventBus.new()
    self.rng     = RNG.new(opts.seed)

    self.scoring = Scoring.new(self.events)
    self.wheel   = WheelS.new(self.events)
    self.shop    = Shop.new(self.events)
    self.choice  = ChoiceS.new(self.events)
    self.meta    = Meta.new(self.events)

    self.state = StateMod.createGameState(self.rng:getSeed())
    if opts.meta then
        for k, v in pairs(opts.meta) do self.state.meta[k] = v end
    end
    return self
end

function GL:_getMods()
    return Effect.compute(self.state.run and self.state.run.relics or {})
end

function GL:_setPhase(p)
    if self.state.phase == p then return end
    self.state.phase = p
    self.events:emit('phase:changed', { phase = p })
end

function GL:_error(msg)
    self.events:emit('error', { message = msg })
    return false
end

-- ─── Lifecycle ────────────────────────────────────────────────

function GL:startRun()
    local ok = { [PHASE.IDLE] = true, [PHASE.GAME_OVER] = true, [PHASE.VICTORY] = true }
    if not ok[self.state.phase] then
        return self:_error('Cannot start run in phase ' .. self.state.phase)
    end

    StateMod.resetUid()
    self.state.run = StateMod.createRunState()
    self:_setPhase(PHASE.IDLE)
    self:_emitRoundPreview()
    self.events:emit('run:started', { seed = self.rng:getSeed() })
    return true
end

-- ─── Spin (RNG-based, for tests / headless) ─────────────────

function GL:spin()
    local ok = { [PHASE.IDLE] = true, [PHASE.SPINNING] = true }
    if not ok[self.state.phase] then
        return self:_error('Cannot spin in phase ' .. self.state.phase)
    end

    local run = self.state.run
    if not run then return self:_error('No active run') end
    if run.ballsLeft <= 0 then
        self.events:emit('error', { message = 'Plus de billes' })
        return nil
    end

    self:_setPhase(PHASE.SPINNING)
    local result = self.wheel:spin(run, self.rng)
    return self:_recordBall(run, result.segmentIndex)
end

-- ─── Resolve ball at specific segment (physics-driven) ──────

function GL:resolveBallAt(segmentIndex)
    local ok = { [PHASE.IDLE] = true, [PHASE.SPINNING] = true }
    if not ok[self.state.phase] then
        return self:_error('Cannot resolve ball in phase ' .. self.state.phase)
    end

    local run = self.state.run
    if not run then return self:_error('No active run') end
    if run.ballsLeft <= 0 then return nil end

    local segment = run.wheel[segmentIndex + 1]  -- 0-based → 1-based
    if not segment then return self:_error('Invalid segment index: ' .. segmentIndex) end

    self:_setPhase(PHASE.SPINNING)
    return self:_recordBall(run, segmentIndex)
end

-- ─── Shared ball resolution ─────────────────────────────────

function GL:_recordBall(run, segmentIndex)
    -- Check special ball
    local specialBall = nil
    if not run._specialBallsFired then run._specialBallsFired = 0 end
    if run._specialBallsFired < #run.specialBalls then
        specialBall = run.specialBalls[run._specialBallsFired + 1]
        run._specialBallsFired = run._specialBallsFired + 1
    end

    local seg, symbol, value = self:_resolveSegment(run, segmentIndex, specialBall)

    local spinResult = {
        segmentIndex = segmentIndex,
        segment = seg,
        symbol = symbol,
        value = value,
        specialBall = specialBall,
    }
    table.insert(run.spinResults, spinResult)
    run.score = run.score + value
    run.ballsLeft = run.ballsLeft - 1

    -- Ticket ball: award tickets equal to pocket number
    if specialBall and specialBall.effect == 'ticket' then
        local ticketsEarned = segmentIndex + 1
        self.state.meta.tickets = self.state.meta.tickets + ticketsEarned
        self.state.meta.totalTickets = (self.state.meta.totalTickets or 0) + ticketsEarned
        spinResult.ticketsEarned = ticketsEarned
        self.events:emit('tickets:earned', { amount = ticketsEarned, total = self.state.meta.tickets })
    end

    -- Splash ball: also score adjacent segments
    if specialBall and specialBall.effect == 'splash' then
        local len = #run.wheel
        local left = ((segmentIndex - 1) + len) % len
        local right = (segmentIndex + 1) % len
        local _, _, vL = self:_resolveSegment(run, left, nil)
        local _, _, vR = self:_resolveSegment(run, right, nil)
        run.score = run.score + vL + vR
        spinResult.splashValue = vL + vR
        spinResult.value = spinResult.value + vL + vR
    end

    local isGold = false
    for _, gp in ipairs(BALANCE.GOLD_POCKETS) do
        if gp == segmentIndex then isGold = true; break end
    end

    self.events:emit('ball:resolved', {
        result = spinResult,
        value = spinResult.value,
        ballsLeft = run.ballsLeft,
        specialBall = specialBall,
        isGold = isGold,
        segmentIndex = segmentIndex,
        score = run.score,
    })

    -- Round end is now driven by Spin scene after full reveal+quota anim
    -- (see Spin:_resolveAll → loop:finishSpinRound)

    return { result = spinResult, value = spinResult.value }
end

-- Called by the Spin scene once reveal sequence + quota deduction anim are done.
function GL:finishSpinRound()
    if self.state.phase ~= PHASE.SPINNING then return false end
    local run = self.state.run
    if not run or run.ballsLeft > 0 then return false end
    self:_endRound()
    return true
end

function GL:_resolveSegment(run, segmentIndex, specialBall)
    local segment = run.wheel[segmentIndex + 1]
    local symbol = segment.symbolId and getSymbol(segment.symbolId) or nil
    local mods = self:_getMods()

    -- Base value: relic override or pocket number (1-indexed)
    local baseVal
    if mods.setBaseValue ~= nil then
        baseVal = mods.setBaseValue
    else
        baseVal = segmentIndex + 1
    end

    -- Even/odd value bonuses from relics
    if baseVal % 2 == 0 then
        baseVal = baseVal + mods.addEven
    else
        baseVal = baseVal + mods.addOdd
    end

    local value = baseVal * segment.weight

    -- Wheel upgrade: value_plus_2
    local upgradeCount = 0
    for _, u in ipairs(run.purchasedUpgrades or {}) do
        if u.effect == 'value_plus_2' then upgradeCount = upgradeCount + 1 end
    end
    if upgradeCount > 0 then value = value + 2 * upgradeCount end

    -- Gold pocket: ×2
    local isGold = false
    for _, gp in ipairs(BALANCE.GOLD_POCKETS) do
        if gp == segmentIndex then isGold = true; break end
    end
    if isGold then value = value * 2 end

    -- Special ball effects
    if specialBall then
        if specialBall.effect == 'double' then
            value = value * 2
        elseif specialBall.effect == 'critical' then
            value = value * 5
        elseif specialBall.effect == 'ticket' then
            value = 0
        end
    end

    return segment, symbol, value
end

function GL:getSegmentDisplayValues()
    local run = self.state.run
    if not run then return {} end
    local mods = self:_getMods()
    local values = {}
    for i, seg in ipairs(run.wheel) do
        local v
        if mods.setBaseValue ~= nil then
            v = mods.setBaseValue
        else
            v = i  -- 1-indexed = segmentIndex+1
        end
        if v % 2 == 0 then v = v + mods.addEven
        else v = v + mods.addOdd end
        values[i] = v
    end
    return values
end

-- ─── End of round → Results ──────────────────────────────────

function GL:_endRound()
    local run = self.state.run
    run._spinInProgress = false

    local result = self.scoring:evaluateRound(run)

    run.lastRoundResult = {
        round = run.round,
        totalWon = result.totalWon,
        quota = result.quota,
        passed = result.passed,
        surplus = result.surplus,
        shopCoins = result.shopCoins,
        ticketsEarned = 0,
    }

    run.shopCurrency = run.shopCurrency + result.shopCoins

    -- Award tickets for passing the round
    if result.passed then
        local earned = BALANCE.TICKETS_PER_ROUND
        self.state.meta.tickets = self.state.meta.tickets + earned
        self.state.meta.totalTickets = (self.state.meta.totalTickets or 0) + earned
        run.lastRoundResult.ticketsEarned = earned
        self.events:emit('tickets:earned', { amount = earned, total = self.state.meta.tickets })
    end

    self:_setPhase(PHASE.RESULTS)
    self.events:emit('round:ended', run.lastRoundResult)

    if not result.passed then
        self:_gameOver()
    end
end

-- ─── Results → Choice ────────────────────────────────────────

function GL:continueFromResults()
    if self.state.phase ~= PHASE.RESULTS then return false end

    local run = self.state.run
    if not run.lastRoundResult.passed then return false end

    if run.round >= BALANCE.ROUNDS_PER_RUN then
        self:_victory()
        return true
    end

    local choices = self.choice:generate(run, self.state.meta, self.rng)
    run.currentChoices = choices

    self:_setPhase(PHASE.CHOICE)
    self.events:emit('choice:presented', { choices = choices })
    return true
end

-- ─── Choice → Shop ───────────────────────────────────────────

function GL:makeChoice(index, targetIndex)
    if self.state.phase ~= PHASE.CHOICE then
        return self:_error('Not in CHOICE phase')
    end

    local run = self.state.run
    local choice = run.currentChoices[index]
    if not choice then
        return self:_error('Invalid choice index')
    end

    local ok = self.choice:apply(run, choice, targetIndex, self.wheel)
    if not ok then return false end

    self.events:emit('choice:made', { choice = choice, index = index })
    self:_openShop()
    return true
end

function GL:skipChoice()
    if self.state.phase ~= PHASE.CHOICE then return false end
    self:_openShop()
    return true
end

function GL:_openShop()
    local run = self.state.run
    run.shopOfferings = self.shop:generateOfferings(run, self.rng)
    run.rerollCount = 0

    self:_setPhase(PHASE.SHOP)
    self.events:emit('shop:opened', { tickets = self.state.meta.tickets, offerings = run.shopOfferings })
end

-- ─── Shop ─────────────────────────────────────────────────────

function GL:shopBuy(slotIndex)
    if self.state.phase ~= PHASE.SHOP then return self:_error('Not in SHOP') end
    return self.shop:buyItem(self.state.run, self.state.meta, slotIndex, self.wheel)
end

function GL:shopReroll()
    if self.state.phase ~= PHASE.SHOP then return self:_error('Not in SHOP') end
    return self.shop:reroll(self.state.run, self.state.meta, self.rng)
end

function GL:endShop()
    if self.state.phase ~= PHASE.SHOP then return false end

    local run = self.state.run
    run.round = run.round + 1

    if run.round > BALANCE.ROUNDS_PER_RUN then
        self:_victory()
        return true
    end

    -- Deduct quota from accumulated score (surplus carries over)
    if run.lastRoundResult and run.lastRoundResult.passed then
        run.score = run.score - run.lastRoundResult.quota
    end

    -- Reset for next round
    run._spinInProgress = false
    run._specialBallsFired = 0
    run.ballsLeft = BALANCE.BALLS_PER_ROUND + #run.specialBalls + (run.genericBallsBought or 0)
    run.spinResults = {}
    run.shopDiscount = 0

    self:_setPhase(PHASE.IDLE)
    self:_emitRoundPreview()
    return true
end

-- ─── End states ───────────────────────────────────────────────

function GL:_gameOver()
    local run = self.state.run
    self:_applyRunStats(run)

    self:_setPhase(PHASE.GAME_OVER)
    self.events:emit('game:over', {
        round = run.round,
        score = run.score,
        quota = getQuota(run.round),
        totalWon = run.lastRoundResult and run.lastRoundResult.totalWon or 0,
    })
end

function GL:_victory()
    local run = self.state.run
    local bonus = BALANCE.TICKETS_BONUS_WIN
    self.state.meta.tickets = self.state.meta.tickets + bonus
    self.state.meta.totalTickets = (self.state.meta.totalTickets or 0) + bonus
    self:_applyRunStats(run)

    self:_setPhase(PHASE.VICTORY)
    self.events:emit('game:won', { round = run.round, tickets = bonus, score = run.score })
end

function GL:_applyRunStats(run)
    self.state.meta.runsCompleted = self.state.meta.runsCompleted + 1
    self.state.meta.bestRound = math.max(self.state.meta.bestRound, run.round)
end

-- ─── Public phase trigger (for UI: IDLE → SPINNING) ─────────

function GL:beginSpinning()
    if self.state.phase ~= PHASE.IDLE then return false end
    if not self.state.run or self.state.run.ballsLeft <= 0 then return false end
    if self.state.run._spinInProgress then return false end
    self.state.run._spinInProgress = true
    self:_setPhase(PHASE.SPINNING)
    return true
end

-- ─── Helpers ──────────────────────────────────────────────────

function GL:_emitRoundPreview()
    local run = self.state.run
    self.events:emit('round:preview', {
        round = run.round,
        quota = getQuota(run.round),
        ballsLeft = run.ballsLeft,
        wheel = run.wheel,
        probabilities = self.wheel:getProbabilities(run.wheel),
        relics = run.relics,
    })
end

return GL

File: legacy/src/core/GameLoop.js
import { EventBus } from './EventBus.js';
import { RNG } from './RNG.js';
import { createGameState, createRunState, PHASE } from './GameState.js';
import { BALANCE, getQuota } from '../data/balance.js';
import { getSymbol } from '../data/symbols.js';
import { WheelSystem } from '../systems/WheelSystem.js';
import { ScoringSystem } from '../systems/ScoringSystem.js';
import { EffectSystem } from '../systems/EffectSystem.js';
import { ChoiceSystem } from '../systems/ChoiceSystem.js';
import { ShopSystem } from '../systems/ShopSystem.js';
import { MetaSystem } from '../systems/MetaSystem.js';

/**
 * Main game orchestrator — owns state, delegates to systems, manages phases.
 * Balls land on segments → value calculated from symbol × weight × modifiers.
 * `spin()` picks randomly (for tests); `resolveBallAt(idx)` accepts physics result.
 */
export class GameLoop {
  constructor(options = {}) {
    this.events = new EventBus();
    this.rng = new RNG(options.seed ?? Date.now());
    this.state = createGameState(this.rng.seed);

    if (options.meta) {
      this.state.meta = { ...this.state.meta, ...options.meta };
    }

    this.wheel = new WheelSystem(this.events);
    this.scoring = new ScoringSystem();
    this.effects = new EffectSystem();
    this.choice = new ChoiceSystem(this.events);
    this.shop = new ShopSystem(this.events);
    this.meta = new MetaSystem(this.events);
  }

  #getMods() {
    return this.effects.compute(this.state.run?.relics ?? []);
  }

  // ─── Lifecycle ───

  startRun() {
    const ok = [PHASE.IDLE, PHASE.GAME_OVER, PHASE.VICTORY];
    if (!ok.includes(this.state.phase)) {
      return this.#error('Cannot start run in phase ' + this.state.phase);
    }

    this.state.run = createRunState();
    this.#setPhase(PHASE.IDLE);
    this.#emitRoundPreview();
    this.events.emit('run:started', { seed: this.rng.seed });
    return true;
  }

  // ─── Spin (RNG-based, for tests / headless) ───

  spin() {
    const ok = [PHASE.IDLE, PHASE.SPINNING];
    if (!ok.includes(this.state.phase)) {
      return this.#error('Cannot spin in phase ' + this.state.phase);
    }

    const run = this.state.run;
    if (!run) return this.#error('No active run');
    if (run.ballsLeft <= 0) {
      this.events.emit('error', { message: 'Plus de billes' });
      return null;
    }

    this.#setPhase(PHASE.SPINNING);
    const result = this.wheel.spin(run, this.rng);
    return this.#recordBall(run, result.segmentIndex);
  }

  // ─── Resolve ball at specific segment (physics-driven) ───

  resolveBallAt(segmentIndex) {
    const ok = [PHASE.IDLE, PHASE.SPINNING];
    if (!ok.includes(this.state.phase)) {
      return this.#error('Cannot resolve ball in phase ' + this.state.phase);
    }

    const run = this.state.run;
    if (!run) return this.#error('No active run');
    if (run.ballsLeft <= 0) return null;

    const segment = run.wheel[segmentIndex];
    if (!segment) return this.#error('Invalid segment index: ' + segmentIndex);

    this.#setPhase(PHASE.SPINNING);
    return this.#recordBall(run, segmentIndex);
  }

  // ─── Shared ball resolution ───

  #recordBall(run, segmentIndex) {
    // Check if this ball is a special ball (special balls fire first)
    // specialBallsFired tracks how many have been used this round (reset each round)
    let specialBall = null;
    if (!run._specialBallsFired) run._specialBallsFired = 0;
    if (run._specialBallsFired < run.specialBalls.length) {
      specialBall = run.specialBalls[run._specialBallsFired];
      run._specialBallsFired++;
    }

    const { segment, symbol, value } = this.#resolveSegment(run, segmentIndex, specialBall);

    const isMiss = BALANCE.MISS_POCKETS.includes(segmentIndex);
    const spinResult = { segmentIndex, segment, symbol, value, specialBall, isMiss };
    run.spinResults.push(spinResult);
    run.score += value;
    run.ballsLeft--;

    // Corruption tick
    run.corruption = Math.min(1, run.corruption + BALANCE.CORRUPTION_PER_SPIN);

    // DDA updates
    const dda = run.dda;
    if (isMiss || value === 0) {
      dda.consecutiveMisses++;
      dda.consecutiveHits = 0;
      dda.frustrationScore = Math.min(1, dda.frustrationScore + 0.15);
      dda.flowScore = Math.max(0, dda.flowScore - 0.2);
    } else {
      dda.consecutiveHits++;
      dda.consecutiveMisses = 0;
      dda.flowScore = Math.min(1, dda.flowScore + 0.15);
      dda.frustrationScore = Math.max(0, dda.frustrationScore - 0.1);
    }
    // Tilt: high when close to quota but failing
    const quota = getQuota(run.round);
    if (run.score >= quota * 0.7 && run.score < quota && run.ballsLeft <= 1) {
      dda.tiltScore = Math.min(1, dda.tiltScore + 0.3);
    } else {
      dda.tiltScore = Math.max(0, dda.tiltScore - 0.05);
    }

    // Ticket ball: award tickets equal to pocket number
    if (specialBall?.effect === 'ticket') {
      const ticketsEarned = segmentIndex + 1;
      this.state.meta.tickets += ticketsEarned;
      this.state.meta.totalTickets = (this.state.meta.totalTickets || 0) + ticketsEarned;
      spinResult.ticketsEarned = ticketsEarned;
      this.events.emit('tickets:earned', { amount: ticketsEarned, total: this.state.meta.tickets });
    }

    // Splash ball: also score adjacent segments
    if (specialBall?.effect === 'splash') {
      const len = run.wheel.length;
      const left = (segmentIndex - 1 + len) % len;
      const right = (segmentIndex + 1) % len;
      const { value: vL } = this.#resolveSegment(run, left, null);
      const { value: vR } = this.#resolveSegment(run, right, null);
      run.score += vL + vR;
      spinResult.splashValue = vL + vR;
      spinResult.value += vL + vR;
    }

    this.events.emit('spin:resolved', {
      result: spinResult,
      value: spinResult.value,
      ballsLeft: run.ballsLeft,
      specialBall,
    });

    if (run.ballsLeft <= 0) {
      this.#endRound();
    }

    return { result: spinResult, value: spinResult.value };
  }

  #resolveSegment(run, segmentIndex, specialBall) {
    const segment = run.wheel[segmentIndex];
    const symbol = segment.symbolId ? getSymbol(segment.symbolId) : null;
    const mods = this.#getMods();

    // Base value: relic override or pocket number (1-indexed)
    let baseVal = mods.setBaseValue !== null ? mods.setBaseValue : (segmentIndex + 1);

    // Corruption penalty: above critical threshold, lose 1 base value
    if (run.corruption > BALANCE.CORRUPTION_CRITICAL) {
      baseVal = Math.max(1, baseVal - 1);
    }

    // Even/odd value bonuses from relics
    if (baseVal % 2 === 0) baseVal += mods.addEven;
    else                   baseVal += mods.addOdd;

    let value = baseVal * segment.weight;

    // Wheel upgrade: value_plus_2
    const upgradeCount = (run.purchasedUpgrades || []).filter(u => u.effect === 'value_plus_2').length;
    if (upgradeCount > 0) value += 2 * upgradeCount;

    // Gold pocket: ×2
    if (BALANCE.GOLD_POCKETS.includes(segmentIndex)) {
      value *= 2;
    }

    // Miss pocket: 0
    if (BALANCE.MISS_POCKETS.includes(segmentIndex)) {
      value = 0;
    }

    // Special ball effects (post-symbol)
    if (specialBall?.effect === 'double') {
      value *= 2;
    } else if (specialBall?.effect === 'critical') {
      value *= 5;
    } else if (specialBall?.effect === 'ticket') {
      value = 0;
    }

    return { segment, symbol, value };
  }

  getSegmentDisplayValues() {
    const run = this.state.run;
    if (!run) return [];
    const mods = this.#getMods();
    return run.wheel.map((seg, i) => {
      let v = mods.setBaseValue !== null ? mods.setBaseValue : (i + 1);
      if (v % 2 === 0) v += mods.addEven;
      else             v += mods.addOdd;
      return v;
    });
  }

  // ─── End of round → Results ───

  #endRound() {
    const run = this.state.run;

    const { totalWon, quota, passed, surplus, shopCoins } = this.scoring.evaluateRound(run);

    run.lastRoundResult = {
      round: run.round,
      totalWon,
      quota,
      passed,
      surplus,
      shopCoins,
    };

    run.shopCurrency += shopCoins;

    if (passed) {
      const earned = BALANCE.TICKETS_PER_ROUND;
      this.state.meta.tickets += earned;
      this.state.meta.totalTickets += earned;
      this.events.emit('tickets:earned', { amount: earned, total: this.state.meta.tickets });

      this.#setPhase(PHASE.RESULTS);
      this.events.emit('round:ended', run.lastRoundResult);
    } else if (!run.freeSpinUsed) {
      // Offer one free spin before game over
      this.#setPhase(PHASE.FREE_SPIN_OFFER);
      this.events.emit('free_spin:offered', { score: totalWon, quota });
    } else {
      this.#setPhase(PHASE.RESULTS);
      this.events.emit('round:ended', run.lastRoundResult);
      this.#gameOver();
    }
  }

  // ─── Free Spin ───

  acceptFreeSpin() {
    if (this.state.phase !== PHASE.FREE_SPIN_OFFER) return false;
    const run = this.state.run;
    run.freeSpinUsed = true;
    run.ballsLeft = 1;
    this.#setPhase(PHASE.IDLE);
    this.events.emit('free_spin:accepted', { ballsLeft: 1 });
    return true;
  }

  declineFreeSpin() {
    if (this.state.phase !== PHASE.FREE_SPIN_OFFER) return false;
    const run = this.state.run;
    run.freeSpinUsed = true;
    this.#setPhase(PHASE.RESULTS);
    this.events.emit('round:ended', run.lastRoundResult);
    this.#gameOver();
    return true;
  }

  // ─── Results → Choice ───

  continueFromResults() {
    if (this.state.phase !== PHASE.RESULTS) return false;

    const run = this.state.run;
    if (!run.lastRoundResult.passed) return false;

    if (run.round >= BALANCE.ROUNDS_PER_RUN) {
      this.#victory();
      return true;
    }

    const choices = this.choice.generate(run, this.state.meta, this.rng);
    run.currentChoices = choices;

    this.#setPhase(PHASE.CHOICE);
    this.events.emit('choice:presented', { choices });
    return true;
  }

  // ─── Choice → Shop ───

  makeChoice(index, targetIndex = null) {
    if (this.state.phase !== PHASE.CHOICE) {
      return this.#error('Not in CHOICE phase');
    }

    const run = this.state.run;
    if (index < 0 || index >= run.currentChoices.length) {
      return this.#error('Invalid choice index');
    }

    const choice = run.currentChoices[index];
    const ok = this.choice.apply(run, choice, targetIndex, this.wheel);
    if (!ok) return false;

    this.events.emit('choice:made', { choice, index });
    this.#openShop();
    return true;
  }

  skipChoice() {
    if (this.state.phase !== PHASE.CHOICE) return false;
    this.#openShop();
    return true;
  }

  #openShop() {
    const run = this.state.run;
    run.shopOfferings = this.shop.generateOfferings(run, this.rng);
    run.rerollCount = 0;

    this.#setPhase(PHASE.SHOP);
    this.events.emit('shop:opened', { tickets: this.state.meta.tickets, offerings: run.shopOfferings });
  }

  // ─── Shop ───

  shopBuyRelic(offeringIndex) {
    return this.shopBuy(offeringIndex);
  }

  shopBuy(slotIndex) {
    if (this.state.phase !== PHASE.SHOP) return this.#error('Not in SHOP');
    return this.shop.buyItem(this.state.run, this.state.meta, slotIndex, this.wheel);
  }

  shopReroll() {
    if (this.state.phase !== PHASE.SHOP) return this.#error('Not in SHOP');
    return this.shop.reroll(this.state.run, this.state.meta, this.rng);
  }

  endShop() {
    if (this.state.phase !== PHASE.SHOP) return false;

    const run = this.state.run;

    run.round++;

    if (run.round > BALANCE.ROUNDS_PER_RUN) {
      this.#victory();
      return true;
    }

    // Deduct quota from accumulated score (surplus carries over to next round)
    if (run.lastRoundResult?.passed) {
      run.score -= run.lastRoundResult.quota;
    }

    // Reset for next round — special balls + generic balls bought carry over
    run._specialBallsFired = 0;
    run.ballsLeft = BALANCE.BALLS_PER_ROUND + run.specialBalls.length + (run.genericBallsBought || 0);
    run.spinResults = [];
    run.shopDiscount = 0;
    run.dda.nearMissesThisRound = 0;

    this.#setPhase(PHASE.IDLE);
    this.#emitRoundPreview();
    return true;
  }

  // ─── End states ───

  #gameOver() {
    const run = this.state.run;
    this.state.meta.lastLostRound = run.round;
    this.#applyRunStats(run);

    this.#setPhase(PHASE.GAME_OVER);
    this.events.emit('game:over', {
      round: run.round,
      score: run.score,
      quota: getQuota(run.round),
      totalWon: run.lastRoundResult?.totalWon ?? 0,
    });
  }

  #victory() {
    const run = this.state.run;
    const bonus = BALANCE.TICKETS_BONUS_WIN;
    this.state.meta.tickets += bonus;
    this.state.meta.totalTickets += bonus;
    this.#applyRunStats(run);

    this.#setPhase(PHASE.VICTORY);
    this.events.emit('game:won', { round: run.round, tickets: bonus, score: run.score });
  }

  #applyRunStats(run) {
    this.state.meta.runsCompleted++;
    this.state.meta.bestRound = Math.max(this.state.meta.bestRound, run.round);
  }

  // ─── Helpers ───

  #setPhase(phase) {
    this.state.phase = phase;
    this.events.emit('phase:changed', { phase });
  }

  #emitRoundPreview() {
    const run = this.state.run;
    this.events.emit('round:preview', {
      round: run.round,
      quota: getQuota(run.round),
      ballsLeft: run.ballsLeft,
      wheel: run.wheel,
      probabilities: this.wheel.getProbabilities(run.wheel),
      relics: run.relics,
    });
  }

  #error(msg) {
    this.events.emit('error', { message: msg });
    return false;
  }
}
